paparazzi/src/main/java/app/cash/paparazzi/internal/apng/ApngVerifier.kt

```diff
@@ -0,0 +1,221 @@
+    var currentDelta = deltaImage
+    repeat(actualDeltasPerFrame) {
+      writer.writeImage(currentDelta)
+      // While adding delta frames for this actual frame, compare the current output video's frame #
+      // with the expected video's delta frame count to decide if we need to prepare the next
+      // expected frame for the delta image diff.
+      if (writer.frameCount % expectedDeltasPerFrame == 0) {
+        currentGoldenFrame = pngReader.readNextFrame()
+        val (expectedFrame, actualFrame) = resize(currentGoldenFrame ?: blankFrame, image)
+        currentDelta = ImageUtils.compareImages(expectedFrame, actualFrame, withErrorText).first
+      }
+    }
+  }
+
+  fun assertFinished() {
+    if (deltaWriter != null) {
+      currentGoldenFrame?.let { lastFrame ->
+        val (expectedFrame, actualFrame) = resize(lastFrame, blankFrame)
+        val (currentDelta, diff) = ImageUtils.compareImages(expectedFrame, actualFrame, withErrorText)
+
+        val times = expectedDeltasPerFrame - (deltaWriter!!.frameCount % expectedDeltasPerFrame)
+        repeat(times) { deltaWriter!!.writeImage(currentDelta) }
+        invalidFrames += diff
+      }
+
+      while (!pngReader.isFinished()) {
+        val (deltaImage, percentDifferent) = ImageUtils.compareImages(
+          goldenImage = pngReader.readNextFrame()!!,
+          image = blankFrame,
+          withText = withErrorText
+        )
+        repeat(expectedDeltasPerFrame) { deltaWriter!!.writeImage(deltaImage) }
+        invalidFrames += percentDifferent
+      }
+    }
+
+    if (deltaWriter != null) {
+      val error = buildString {
+        if (invalidFrames.size >= 1) {
+          appendLine(
+            "${invalidFrames.size} frames differed by more than %.1f%%".format(maxPercentDifference)
+          )
+        }
+        if (pngReader.getFps() != fps) {
+          appendLine("Mismatched video fps expected: ${pngReader.getFps()} actual: $fps")
+        }
+        if (pngReader.frameCount != frameCount) {
+          appendLine("Mismatched frame count expected: ${pngReader.frameCount} actual: $frameCount")
+        }
+        appendLine(" - see details in file://$deltaFilePath\n")
+      }
+
+      if (error.isNotEmpty()) {
+        fail(error)
+      }
+    }
+  }
+
+  /**
+   * Initializes the writer that generates delta images for all frames excluding the current
+   * frame which is about to be written by the [verifyFrame] method.
+   */
+  private fun ApngReader.initializeWriter(): ApngWriter {
+    return ApngWriter(deltaFilePath, commonFrameRate, fileSystem).apply {
+      val currentFrameNumber = frameNumber - 1
+      reset()
+      repeat(currentFrameNumber) {
+        val nextFrame = readNextFrame() ?: blankFrame
+        val (deltaImage) = ImageUtils.compareImages(
+          goldenImage = nextFrame,
+          image = nextFrame,
+          withText = withErrorText
+        )
+
+        writeImage(deltaImage)
+      }
+      readNextFrame()
+    }
+  }
```

Comment:
It seems like the hard part is discerning reader from writer. I think this would strike a happy medium

```
  private fun ApngReader.initializeWriter(): ApngWriter {
    val currentFrameNumber = frameNumber - 1
    reset()

    return ApngWriter(deltaFilePath, commonFrameRate, fileSystem).apply {
      repeat(currentFrameNumber) {
        val nextFrame = readNextFrame() ?: blankFrame
        val (deltaImage) = ImageUtils.compareImages(
          expected = nextFrame,
          actual = nextFrame,
          withText = withErrorText
        )

        writeImage(deltaImage)
      }
      readNextFrame()
    }
  }
  ```